packages/jsonata-paths-extractor/src/index.ts

```diff
@@ -0,0 +1,289 @@
+import * as jsonata from 'jsonata';
+interface PathContext {
+  pathNode: jsonata.ExprNode;
+  pathVariablesInScope: Record<string, PathContext>;
+  pathId: Symbol;
+  parentPathId?: Symbol;
+}
+const RootContextId = Symbol('ROOT');
+
+const UnknownContextId = Symbol('UNKNOWN');
+
+const createUnknownContext = (
+  ast: jsonata.ExprNode,
+  pathVariablesInScope: Record<string, PathContext>
+) => {
+  return {
+    pathNode: ast,
+    pathVariablesInScope: { ...pathVariablesInScope },
+    pathId: UnknownContextId,
+  };
+};
+
+const createRootContext = (ast: jsonata.ExprNode) => {
+  return {
+    pathNode: ast,
+    pathVariablesInScope: {},
+    pathId: RootContextId,
+  };
+};
+
+const reduceToPathString = (
+  pathContext: PathContext,
+  stopAtPosition?: number
+): string => {
+  return pathContext.pathNode
+    .steps!.filter(step =>
+      typeof stopAtPosition === 'number'
+        ? step.position! < stopAtPosition
```

Comment:
Another non-null assertion, can `step.position!` definitely not fail?

Reply:
refactored, :+1: